plutus-ir/src/Language/PlutusIR/Parser.hs

```diff
@@ -0,0 +1,344 @@
+{-# LANGUAGE FlexibleContexts #-}
+module Language.PlutusIR.Parser where
+
+import           Prelude                      hiding (fail)
+
+import           Control.Applicative          hiding (many, some)
+import           Control.Monad.State          hiding (fail)
+
+import qualified Language.PlutusCore          as PLC
+import qualified Language.PlutusCore.Constant as PLC
+import           Language.PlutusIR            as PIR
+import           Text.Megaparsec              hiding (State, parse)
+
+import           Data.ByteString.Internal     (c2w)
+import qualified Data.ByteString.Lazy         as BS
+import           Data.Char
+import           Data.Foldable
+import qualified Data.Map                     as M
+import qualified Data.Text                    as T
+import           Data.Word
+import           GHC.Natural
+
+import           Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer   as Lex
+
+data ParserState = ParserState { identifiers :: M.Map String PLC.Unique, counter :: PLC.Unique }
+    deriving (Show)
+
+data PIRParseError = Overflow Natural Integer
+                   | UnexpectedKeyword String
+                   | InternalError String
+                   deriving (Eq, Ord, Show)
+
+instance ShowErrorComponent PIRParseError where
+    showErrorComponent (Overflow sz i) = "Integer overflow: " ++ show i ++ " does not fit in " ++ show sz ++ " bytes"
+    showErrorComponent (UnexpectedKeyword kw) = "Keyword " ++ kw ++ " used as identifier"
+    showErrorComponent (InternalError cause) = "Internal error: " ++ cause
+
+initial :: ParserState
+initial = ParserState M.empty (PLC.Unique 0)
+
+intern :: MonadState ParserState m => String -> m PLC.Unique
+intern n = do
+    st <- get
+    case M.lookup n (identifiers st) of
+        Just u -> return u
+        Nothing -> do
+            let counter' = PLC.Unique $ (PLC.unUnique $ counter st) + 1
+                identifiers' = M.insert n counter' $ identifiers st
+            put $ ParserState identifiers' counter'
+            return counter'
+
+type Parser = ParsecT PIRParseError String (State ParserState)
+
+parse :: String -> Parser a -> String -> Either (ParseError Char PIRParseError) a
+parse filename p s = evalState (runParserT p filename s) initial
```

Comment:
I find that the monad transformer functions have the arguments in a bit of annoying order for things like this. I would probably write
```
parse = flip evalState initial  . runParserT
```
(and flip the first two arguments of the function as well)